File: Code/src/Aufgabe3.java
import java.util.Arrays;

public final class Aufgabe3 {
    
    
    static final class IntStack {
        private int[] stack = new int[10];
        private int size;
        
        final boolean isEmpty() {
            return this.size == 0;
        }
        
        
        final void push(int element) {
            if (this.size >= this.stack.length) {
                this.stack = Arrays.copyOf(this.stack, this.stack.length * 2);
            }
            
            this.stack[this.size] = element;
            this.size++;
        }
        
        final int peek() {
            return this.stack[this.size - 1];
        }
        
        final int pop() {
            int result = this.stack[this.size - 1];
            this.size--;
            return result;
        }
    }
    
    private static int f(int n, int m) {
        if (n == 0) return m + 1;
        else if (m == 0) return f(n - 1, 1);
        else return f(n - 1, f(n, m - 1));
    }
    
    private static int itF(int paramN, int paramM) {
        IntStack stack = new IntStack();
        stack.push(paramN);
        stack.push(paramM);
        
        while (true) {
            int m = stack.pop();
            int n = stack.pop();
            
            if (n == 0) {
                int result = m + 1;
                if (stack.isEmpty()) {
                    return result;
                }
                
                stack.push(result);
            } else if (m == 0) {
                stack.push(n - 1);
                stack.push(1);
            } else {
                stack.push(n - 1);
                stack.push(n);
                stack.push(m - 1);
            }
        }
    }
    
    public static void main(String[] args) {
        for (int i = 0; i <= 20; i++) {
            System.out.println(f(3, i));
        }
        
    }
}

File: Code/src/Aufgabe3.kt
@Suppress("NON_TAIL_RECURSIVE_CALL")
tailrec fun f(n: Int, m: Int): Int =
    when {
        n == 0 -> m + 1
        m == 0 -> f(n - 1, 1)
        else -> f(n - 1, f(n, m - 1))
    }

class LongStack {
    private var stack = LongArray(10)

    private var size = 0

    fun isEmpty() = size == 0

    fun push(element: Long) {
        if (size >= stack.size) {
            stack = stack.copyOf(stack.size * 2)
        }
        stack[size] = element
        size += 1
    }

    fun peek(): Long = stack[size - 1]
    fun pop() = stack[size -1].also { size -= 1 }
}

@Suppress("NAME_SHADOWING")
fun itF(n: Long, m: Long): Long {
    val stack = LongStack()
    stack.push(n)
    stack.push(m)

    while (true) {
        val m = stack.pop()
        val n = stack.pop()

        when {
            n == 0L -> {
                val result = m + 1
                if (stack.isEmpty()) {
                    return result
                }
                // set new m
                stack.push(result)
            }
            m == 0L -> {
                stack.push(n - 1) // new n
                stack.push(1)     // new m
            }
            else -> {
                // current step
                stack.push(n - 1)

                // recursion for calculating parameter m
                stack.push(n)
                stack.push(m - 1)
            }
        }
    }
}

fun main(args: Array<String>) {
    // f(0, m) ist trivial
    for (i in 2L..30L) {
        println(itF(4, i))
    }


//    println(itF(4, 1)) // 65533
//    println(itF(5, 0)) // 65533

}